### PR TITLE
command and docs for async input

### PR DESCRIPTION
**Pull request by pivnicek**
backend PR keboola/provisioning-router#169, the main issue is keboola/internal#111

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -412,3 +412,27 @@ finished, the result will be contained in the `result` node.
             "status": "waiting"
         }
 
+## Input data Async [/async/docker/{credentials_id}/input]
```
Comment (ondrejhlavacek) [inline apiary.apib]: In Storage API we use the `load` verb (https://keboola.docs.apiary.io/#reference/workspaces/load-data/load-data)

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public function setUp()
+    {
+        $this->client = new Client("docker", PROVISIONING_API_TOKEN, "ProvisioningApiTest", PROVISIONING_API_URL, SYRUP_QUEUE_URL);
+    }
+
+    public function testInputData()
+    {
+        $result = $this->client->getCredentialsAsync("jupyter");
+
+        $response = $this->client->loadData($result['id'], 'input: {
```
Comment (ondrejhlavacek) [inline tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: I'm having trouble understanding the payload. It's not a PHP array, it's not a valid JSON, what is it and how come it works this way?

File: apiary.apib [hunk 3]
```diff
@@ -412,3 +412,27 @@ finished, the result will be contained in the `result` node.
...
+
++ Parameters
+    + credentials_id(required, int) - ID of the credentials
+
+### Input Data Async [POST]
+Asynchronously load data to the dicker sandbox
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply apiary.apib]: lmao :-D

File: apiary.apib (hunk 1 above)
Comment (pivnicek) [reply apiary.apib]: yeah, I have the methods as `load` and `unload` and the endpoints as `input` and `output`, as per discussion in the internal issue.  I can change it, but its a bit of a pita.

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php (hunk 2 above)
Comment (pivnicek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: it's standard input mapping https://github.com/keboola/input-mapping

File: apiary.apib (hunk 1 above)
Comment (ondrejhlavacek) [reply apiary.apib]: where's the issue? no link and no description of this PR :-(

Comment (pivnicek) [reply apiary.apib]: sorry! I should have linked the backend PR https://github.com/keboola/provisioning-router/pull/169, the main issue is https://github.com/keboola/internal/issues/111

Comment (ondrejhlavacek) [reply apiary.apib]: you can still update it now :-)

Comment (pivnicek) [reply apiary.apib]: ok, I will if @odinuv agrees :-)

Comment (ondrejhlavacek) [reply apiary.apib]: i mean the issue description :-)

Comment (pivnicek) [reply apiary.apib]: haha :-) updated

**Comment by odinuv** [comment]
i have no strong opinion about input / load - it doesn't have to be the same as in connection, but to me - input is something you setup at the beginning, so load would seem slightly more suitable. On the other hand - unload to me evokes that the contents are removed in the source

**Comment by pivnicek** [comment]
ok, I also don't feel strongly about it.  I'd rather keep it as `input` and `output` because a) that's how it is right now :) and b) it has nice symmetry with input-mapping and output-mapping imho

**Comment by ondrejhlavacek** [comment]
If the API verb is `input`, then the root node `input` seems duplicate or redundant (but in fact it is notú.

**Comment by ondrejhlavacek** [comment]
for the same of DX i'd try to keep the api endpoints/verbs the same across the whole platform. In the end this will all be a part of analytical workspaces. And if `unload` seems weird we can easily change it to something else.

**Comment by pivnicek** [comment]
we can always alias it later if needed.  we can change it when we make new provisioning too.

**Comment by odinuv** [comment]
also - how are the client tests run?

**Comment by pivnicek** [comment]
they are run in provisioning-router on any tag build https://github.com/keboola/provisioning-router/blob/master/.travis.yml#L66

tagged builds deploy to testing so that's why I tagged that build (which failed :-( ), so then I can test versus testing stack

**Comment by odinuv** [comment]
but that it is testing provisioning-client master, doesn't it ?

**Comment by pivnicek** [comment]
yes.

I run the tests locally with `docker-compose run --rm php ./vendor/bin/phpunit`  with my env set to suryp testing env

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php (hunk 2 above)
Comment (pivnicek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: sorry, that was all messed up :-(.  it should be better now

File: apiary.apib [hunk 4]
```diff
@@ -412,3 +412,27 @@ finished, the result will be contained in the `result` node.
...
+Asynchronously load data to the docker sandbox
+
+#### Attributes
++ input (array) - sandbox input mapping containing tables and files preloaded into the sandbox. Definition follows the [common schema](https://github.com/keboola/docker-bundle/blob/master/Resources/schemas/configuration.json#L30)
+
++ Request
+    + Headers
+
+            X-StorageApi-Token: 123456
+
```
Comment (ondrejhlavacek) [inline apiary.apib]: missing request body

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php (hunk 2 above)
Comment (ondrejhlavacek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: I still don't get it. But the request body is missing in the docs, so I cannot verify that the payload is correct or not.

File: apiary.apib [hunk 5]
```diff
@@ -412,3 +412,27 @@ finished, the result will be contained in the `result` node.
             "status": "waiting"
         }
 
+## Input data Async [/async/docker/{credentials_id}/input]
+
++ Parameters
+    + credentials_id(required, int) - ID of the credentials
+
+### Input Data Async [POST]
+Asynchronously load data to the docker sandbox
```
Comment (ondrejhlavacek) [inline apiary.apib]:
```suggestion
Asynchronously load data to the Docker sandbox
```

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                tables: [
+                    [
+                        source: "in.c-sandbox.test",
+                        destination => "source.csv"
+                    ]
+                ]
+            }');
+            $this->fail('incorrect message body should fail');
+        } catch (\Exception $e) {
+            echo $e->getMessage() . " error code " . $e->getCode();
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: yes, but this is on an ancient phpunit version

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php (hunk 2 above)
Comment (pivnicek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: your wish is my command :-) https://github.com/keboola/provisioning-client/pull/31

File: tests/Keboola/Provisioning/Client/DataLoaderApiTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: the problem is that the `  } catch (\Exception $e) {` catches the exception thrown by this->fail

[the remark this answers is not in the document]
Comment (pivnicek) [reply tests/Keboola/Provisioning/Client/DataLoaderApiTest.php]: doh!

**Comment by pivnicek** [comment]
@odinuv I believe this is ok now, but I can't run it at the same time as the provisioning is running because some of the tests that count number of tasks can fail (on router).  So I'lll re-run when that is done.

**Comment by odinuv** [comment]
@pivnicek you changed the testing project https://travis-ci.com/keboola/provisioning-client/builds/138281519 ?

**Comment by pivnicek** [comment]
Yes, I broke the 572 with concurrent tests and the test bucket is broken now, any operation on it results in a connection app errror.  So I tried switching test projects, which was ok but the other project, 580, doesn't have redshift.  I guess I have 3 options now, 1) fix the broken bucket in 572, 2) add RS to 580, 3) move the tests to keboola-testing.  What do you think @odinuv ?

**Comment by odinuv** [comment]
imho best: create a new project just for this and add redshift there

**Comment by pivnicek** [comment]
So this appears to be passing on the new project finally 🎉 @odinuv

**Comment by pivnicek** [comment]
@ondrejhlavacek pokud budes mit chvilicku muzes na tohle dat oko?

File: apiary.apib (hunk 5 above)
Comment (ondrejhlavacek) [reply apiary.apib]: ping here

**Comment by pivnicek** [comment]
Yeah, the error handling is tested internally.  I may add more of these tests in a later release when I add the other endpoints.

File: apiary.apib [hunk 7]
```diff
@@ -412,3 +412,45 @@ finished, the result will be contained in the `result` node.
...
+
++ Request
+    + Headers
+
+            X-StorageApi-Token: 123456
+
+    + Body
+
+            {
+                "input": {
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply apiary.apib]: @odinuv nene, to dela ten client `loadData` command.  Tady: https://github.com/keboola/provisioning-client/pull/29/files#diff-0d62bbb74d1a1fb1fa88968420f9a924R246

[the remark this answers is not in the document]
Comment (odinuv) [reply apiary.apib]: nene means "yes, shouldn't be here" or "no, it should be here" ? :) 

btw, i have no idea how this is posted 23 mins ago, i definitely wrote it yesterday :/

[the remark this answers is not in the document]
Comment (pivnicek) [reply apiary.apib]: :-) . nene means that the documentation is correct.  The php client unload function adds the input node, so it's not directly visible in the test method

**Pull request merged**